src/shogun/machine/TypedMachine.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Shubham Shukla
+ */
+
+#ifndef _TYPEDMACHINE_H__
+#define _TYPEDMACHINE_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/base/progress.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/LinearMachine.h>
+
+namespace shogun
+{
+	class CFeatures;
+	class CLabels;
+	
+	template <class P, class T>
+	class CTypedMachine : public T
+	{
+	public:
+		/** Default constructor */
+		CTypedMachine() : T()
+		{
+		}
+
+		virtual ~CTypedMachine()
+		{
+		}
+
+	protected:
+		virtual bool train_machine(CFeatures* data = NULL)
+		{
+				if (!data)
+				{	
+					REQUIRE(this->features, "No features provided.\n")	
+					REQUIRE(this->features->get_feature_class() == C_DENSE,	
+					"Feature-class (%d) must be of type C_DENSE (%d)\n", this->features->get_feature_class(), C_DENSE)	
+					
+					data = this->features;	
+				}	
+		else	
+			REQUIRE(data->get_feature_class() == C_DENSE,	
+				"Feature-class must be of type C_DENSE (%d)\n", data->get_feature_class(), C_DENSE)	
+		
+			REQUIRE(data->get_num_vectors() == this->m_labels->get_num_labels(), "Number of training vectors (%d) does not match number of labels (%d)\n"	
+		, data->get_num_vectors(), this->m_labels->get_num_labels())
+			
+			//check for type of CFeatures, then call the appropriate template method
+			if(data->get_feature_type() == F_DREAL)
+				return this->template as<P>()->template train_machine_templated<float64_t>(data->as<CDenseFeatures<float64_t>>());
+			else if(data->get_feature_type() == F_SHORTREAL)
+				return this->template as<P>()->template train_machine_templated<float32_t>(data->as<CDenseFeatures<float32_t>>());
```

Comment:
wow didnt expect we need such things, but yes makes sense if i think about it

Reply:
yes this is a bit tricky

Reply:
I mean `this->template as<P>` might go into some variable.

Reply:
ah i see :)

Reply:
could you paste the compile error that happens when `train_machine_templated` does not exist in CLARS? Curious how it looks

Reply:
`/home/atom/shogun/src/shogun/machine/TypedMachine.h:49:98: error: ‘class shogun::CLDA’ has no member named ‘train_machine_templated’
     return obj->template train_machine_templated<float64_t>(data->as<CDenseFeatures<float64_t>>());` 
also here ```obj = this->as<P>()```

Reply:
this is actually pretty good ...
Just need to make sure in the docs of the CRTP class that we mention that implementing this method is mandatory